Frameworks/CoreText/CTParagraphStyle.mm

```diff
@@ -17,235 +17,182 @@
-                        auto locationToWrite = reinterpret_cast<char*>(property) + offsetForValue;
-                        *reinterpret_cast<CTWritingDirection*>(locationToWrite) =
-                            *reinterpret_cast<const CTWritingDirection*>(settings[i].value);
-                    } else if (settings[i].spec == kCTParagraphStyleSpecifierAlignment) {
-                        auto property = reinterpret_cast<_CTParagraphStyleCTTextAlignmentProperty*>(structContainingData);
-
-                        size_t offsetForValue = offsetof(_CTParagraphStyleCTTextAlignmentProperty, _value);
-                        auto locationToWrite = reinterpret_cast<char*>(property) + offsetForValue;
-                        *reinterpret_cast<CTTextAlignment*>(locationToWrite) = *reinterpret_cast<const CTTextAlignment*>(settings[i].value);
-
-                    } else if (settings[i].spec == kCTParagraphStyleSpecifierLineBreakMode) {
-                        auto property = reinterpret_cast<_CTParagraphStyleCTLineBreakModeProperty*>(structContainingData);
-
-                        size_t offsetForValue = offsetof(_CTParagraphStyleCTLineBreakModeProperty, _value);
-                        auto locationToWrite = reinterpret_cast<char*>(property) + offsetForValue;
-                        *reinterpret_cast<CTLineBreakMode*>(locationToWrite) = *reinterpret_cast<const CTLineBreakMode*>(settings[i].value);
-                    }
-                }
-            }
+            style->SetValue(settings[i]);
         }
     }
 
-    return static_cast<CTParagraphStyleRef>(ret);
+    return style;
 }
 
 /**
  @Status Caveat
- @Notes CTParagraphStyleSpecifier kCTParagraphStyleSpecifierTabStops and kCTParagraphStyleSpecifierCount not supported
+ @Notes CTParagraphStyleSpecifier kCTParagraphStyleSpecifierTabStops
 */
 CTParagraphStyleRef CTParagraphStyleCreateCopy(CTParagraphStyleRef paragraphStyle) {
-    if (!paragraphStyle) {
-        return nil;
-    }
-
-    _CTParagraphStyle* ret = [static_cast<_CTParagraphStyle*>(paragraphStyle) copy];
-    return static_cast<CTParagraphStyleRef>(ret);
+    return paragraphStyle ? __CTParagraphStyle::CreateInstance(nullptr, *paragraphStyle) : nullptr;
 }
 
 /**
  @Status Caveat
- @Notes CTParagraphStyleSpecifier kCTParagraphStyleSpecifierTabStops and kCTParagraphStyleSpecifierCount not supported
+ @Notes CTParagraphStyleSpecifier kCTParagraphStyleSpecifierTabStops
 */
 bool CTParagraphStyleGetValueForSpecifier(CTParagraphStyleRef paragraphStyle,
                                           CTParagraphStyleSpecifier spec,
                                           size_t valueBufferSize,
                                           void* valueBuffer) {
-    if (!paragraphStyle || !valueBuffer) {
-        return false;
+    RETURN_FALSE_IF(paragraphStyle == nullptr);
+    if (paragraphStyle->GetValue(spec, valueBufferSize, valueBuffer)) {
```

Comment:
in fact, right now you will bomb out if `!valueBuffer && validSpecifier(spec)`

---
In reply to: [87272039](https://github.com/Microsoft/WinObjC/pull/1343#discussion_r87272039) [](ancestors = 87272039)